### PR TITLE
Crashreporter

### PR DESCRIPTION
**Pull request by mkestler-mozilla**
Initial code for crash reporting on windows.

### CODE REVIEW

**Comment by flodolo** [comment]
@mkestler-mozilla 
I'm currently getting an email for every commit, so I'd like to unsubscribe ;-)

You're currently changing the whole `translations/strings.yaml`, that should definitely be fixed. Do you plan to have strings in this PR? If so, could you flag me again for review when this is ready?

File: src/crashreporter/crashui.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  window->raise();
+  window->requestActivate();
+}
+
+Q_INVOKABLE void CrashUI::sendReport() {
+  auto window = QmlEngineHolder::instance()->window();
+  window->hide();
+  emit startUpload();
+
+  return Q_INVOKABLE void();
```
Comment (bakulf) [inline src/crashreporter/crashui.cpp:90]: Why this? Can we simply remove this line?

File: src/crashreporter/crashserverclientfactory.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+#include <QObject>
+
+#ifdef MVPN_WINDOWS
+#include "platforms/windows/windowscrashclient.h"
+#endif
+
+using namespace std;
+
+CrashServerClientFactory::CrashServerClientFactory()
```
Comment (bakulf) [inline src/crashreporter/crashserverclientfactory.cpp:15]: Can we merge the server and the client factories?

File: src/crashreporter/crashui/main.qml [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            source: "qrc:/crashresources/Warning.svg"
+            anchors.left: parent.left
+            anchors.right: parent.right
+            horizontalAlignment: Image.AlignHCenter
+            fillMode: Image.PreserveAspectFit
+        }
+          VPNSubtitle {
+              id: mainHeading
+
+              text: qsTrId("vpn.crashreporter.mainHeading")
```
Comment (bakulf) [inline src/crashreporter/crashui/main.qml:82]: This is new string. Use string.yaml

File: src/crashreporter/crashui/main.qml [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+              horizontalAlignment: Text.AlignHCenter
+              //anchors.fill: parent
+          }
+          VPNTextBlock{
+              id: description
+              anchors.left: parent.left
+              anchors.right: parent.right
+              horizontalAlignment: Text.AlignHCenter
+              text: qsTrId("vpn.crashreporter.description")
```
Comment (bakulf) [inline src/crashreporter/crashui/main.qml:91]: Same here.

File: src/crashreporter/crashuploader.cpp [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    logger.error() << "Unable to open file: " << file;
+    nextUpload();
+    return;
+  }
+  m_retries++;
+  if (m_retries > MAX_RETRIES) {
+    nextUpload();
+    return;
+  }
+  QString fileName = file.split("\\").last();
```
Comment (bakulf) [inline src/crashreporter/crashuploader.cpp:63]: Use QFileInfo. It's cross platform

File: src/crashreporter/crashuploader.cpp [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  QByteArray dumpContent = dump.readAll();
+
+  QHttpMultiPart* multipart = new QHttpMultiPart(this);
+  multipart->setBoundary(QString(BOUNDARY).toLocal8Bit());
```
Comment (bakulf) [inline src/crashreporter/crashuploader.cpp:67]: Why a fix string? This can be a random value in theory

File: src/crashreporter/crashuploader.cpp [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+  QHttpPart versionPart;
+  versionPart.setHeader(QNetworkRequest::ContentDispositionHeader,
+                        "form-data; name=\"Version\"");
+  versionPart.setBody(APP_VERSION);
+  multipart->append(namePart);
+  multipart->append(formPart);
+  multipart->append(versionPart);
+
+  QString urlStr = SettingsHolder::instance()->stagingServer()
```
Comment (bakulf) [inline src/crashreporter/crashuploader.cpp:91]: Constants::isProduction

File: src/crashreporter/crashuploader.cpp [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+  request.setHeader(QNetworkRequest::UserAgentHeader, "mozillaVPN");
+  // Qt changes the content type automatically to multipart/mixed.  Socorro
+  // doesn't support that, so we have to set it ourselves and create a custom
+  // bounday.
+  request.setHeader(QNetworkRequest::ContentTypeHeader,
+                    QString("multipart/form-data; boundary=%1").arg(BOUNDARY));
+  auto reply = m_network->post(request, multipart);
+
+  connect(reply, &QNetworkReply::sslErrors, [](QList<QSslError> errors) {
+    cout << "SSL Errors: " << endl;
```
Comment (bakulf) [inline src/crashreporter/crashuploader.cpp:110]: Logger.error

File: src/crashreporter/crashuploader.cpp [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    for (auto err : errors) {
+      std::cout << err.errorString().toStdString() << std::endl;
+    }
+  });
+  QCoreApplication::processEvents(QEventLoop::ExcludeUserInputEvents);
```
Comment (bakulf) [inline src/crashreporter/crashuploader.cpp:115]: Why this? You can use signals and slots to know when the upload is completed and make the code async.

File: src/imageproviderfactory.cpp [hunk 10]
```diff
@@ -0,0 +1,31 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "imageproviderfactory.h"
```
Comment (bakulf) [inline src/imageproviderfactory.cpp:5]: I like this cleanup a lot! maybe i can land independently

File: src/main.cpp [hunk 11]
```diff
@@ -16,17 +30,40 @@ int main(int argc, char* argv[]) {
   Q_UNUSED(leakDetector);
 #  ifdef MVPN_WINDOWS
   // Allocate a console to view log output in debug mode on windows
-  if (AllocConsole()) {
+  /* if (AllocConsole()) {
```
Comment (bakulf) [inline src/main.cpp:33]: ?

File: src/main.cpp [hunk 12]
```diff
@@ -16,17 +30,40 @@ int main(int argc, char* argv[]) {
...
-  }
+
+}*/
 #  endif
 
+#endif
+
+#ifdef MVPN_WINDOWS
+  if (argc > 1) {
+    for (int i = 1; i < argc; i++) {
```
Comment (bakulf) [inline src/main.cpp:48]: Can we use the command line parser?

**Review by bakulf — CHANGES_REQUESTED** [review]
The PR removes a lot of unrelated files. But the core is really promising!

**Comment by mkestler-mozilla** [comment]
Merge from main caused a lot of unrelated issues.  Closing to cleanup.

**Pull request closed without merging**